hal/src/main/java/edu/wpi/first/hal/HALReporter.java

```diff
@@ -0,0 +1,90 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+package edu.wpi.first.hal;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class HALReporter {
+  @SuppressWarnings("MemberName")
+  static class ReportStore {
+    public int resource;
+    public int instanceNumber;
+    public int context;
+    public String feature;
+
+    ReportStore(int resource, int instanceNumber, int context, String feature) {
+      this.resource = resource;
+      this.instanceNumber = instanceNumber;
+      this.context = context;
+      this.feature = feature;
+    }
+  }
+
+  private static final List<ReportStore> storeList = new ArrayList<>();
+
+  private static final Object storeLock = new Object();
+  private static boolean allowDirectWrite = false;
+
+  /**
+   * Write all caches reports to the low level API.
+   *
+   * <p>Not for public use, called from WPILibJ.
+   */
+  public static void writeReports() {
+    synchronized (storeLock) {
+      if (allowDirectWrite) {
+        return;
+      }
+      allowDirectWrite = true;
+    }
+    for (ReportStore r : storeList) {
+      HALReporter.report(r.resource, r.instanceNumber, r.context, r.feature);
+    }
+  }
+
+  /**
+   * Report a usage.
+   *
+   * @param resource resource
+   * @param instanceNumber instanceNumber
+   */
+  public static void report(int resource, int instanceNumber) {
+    report(resource, instanceNumber, 0, "");
+  }
+
+  /**
+   * Report a usage.
+   *
+   * @param resource resource
+   * @param instanceNumber instanceNumber
+   * @param context context
+   */
+  public static void report(int resource, int instanceNumber, int context) {
+    report(resource, instanceNumber, context, "");
+  }
+
+  /**
+   * Report a usage.
+   *
+   * @param resource resource
+   * @param instanceNumber instanceNumber
+   * @param context context
+   * @param feature feature
+   * @return
+   */
+  public static int report(int resource, int instanceNumber, int context, String feature) {
+    synchronized (storeLock) {
+      if (!allowDirectWrite) {
+        storeList.add(new ReportStore(resource, instanceNumber, context, feature));
+        return 0;
+      }
+    }
+    return HALReporter.report(resource, instanceNumber, context, feature);
```

Comment:
Is this meant to be HAL.report instead of a recursive call?

Reply:
Yes. I do need to double check that the JIT doesn't call the static initializer as well. According to the Java spec it can't, but need to double check.